corehq/apps/users/models.py

```diff
@@ -1050,6 +1052,31 @@ def wrap(cls, data, should_save=False):
             del data["organizations"]
             should_save = True
 
+        if "commcare_project" in data.get("user_data", {}):
+            del data["user_data"]["commcare_project"]
+            should_save = True
+
+        if "commcare_location_id" in data.get("user_data", {}):
+            del data["user_data"]["commcare_location_id"]
+            should_save = True
+
+        if "commcare_location_ids" in data.get("user_data", {}):
+            del data["user_data"]["commcare_location_ids"]
+            should_save = True
+
+        if "commcare_primary_case_sharing_id" in data.get("user_data", {}):
+            del data["user_data"]["commcare_primary_case_sharing_id"]
+            should_save = True
+
+        if "commtrack-supply-point" in data.get("user_data", {}):
+            data['commtrack_supply_point'] = data["user_data"]["commtrack-supply-point"]
+            del data["user_data"]["commtrack-supply-point"]
+            should_save = True
+
+        # Wrapping a doc that comes out of ES shouldn't be saved here, even if deprecated keys are present
+        if any(es_param in data for es_param in ["__group_ids", "__group_names", "user_data_es"]):
+            should_save = False
```

Comment:
do you mean instead of doing this lazily, I should do it in a migration? I think that's probably preferable now, good idea.

Reply:
Yeah, or doing it lazily up front for an easier roll-out, _then_ doing the migration.  That can be useful if the migration might take a while and isn't backwards compatible, so you don't get errors in the transition.  I haven't really thought through which would be better here, was just flagging that it's of course preferable to not do permanent lazy migrations.

Reply:
I'm in favor of doing real migrations, just want to make sure that if you're considering a real migration, you're accounting for the number of users on CAS.

Reply:
Yeah, I'm also a little worried that this lazy migration is also going to add a ton of writes to couch right after this change is merged (one for every login, although maybe that isn't an issue?). 

How have we handled user migrations on CAS other times? 

Neither migration is totally necessary for this feature to work (as any stale values will be ignored when the restore is generated), but we would end up with a very confusing set of data if we don't migrate.

Reply:
I don't know if we've ever run a script across all couch users on CAS. @sravfeyn @snopoke we're talking about doing a migration that would remove a couple of user data keys from couch users. Do you have any wisdom to share?

Reply:
If the lazy migration is not necessary then I'd rather have it removed so that we can control when the user updates happen. On CAS this is particularly important since updating every user on their first request after this is merged will likely bring CouchDB to it's knees.

Other than that I don't see an issue with doing an off-deploy migration. It may take a while on ICDS but we can do it off peak.